doc/developer/design/20230607_shift_status_updates.md

```diff
@@ -0,0 +1,150 @@
+
+When dropping a replica, the storage controller will need to retrieve all source/sink dataflows associated with the given storage instance. To achieve this, we can keep a mapping of storage instance id to active ingestions/exports on that instance as part of the storage controller's state. Upon creating collections/exports we can update the map and upon dropping a storage instance, we remove the corresponding entry in the map. The storage controller will only write statuses upon receiving a `StorageResponse::StatusUpdates` response iff the specified `GlobalId` exists in some active mapping. This allows us to ignore any late status updates that are sent back from storage workers.
+
+In order to support the future in which storage clusters may have multiple replicas, we will need to update the schemas for the `mz_{source|sink}_status_history` relations to include a `replica_id`. This is necessary for the reason that dropping a replica in a storage cluster with multiple replicas shouldn't set the status of a source, for example, to `paused`. Isolating status updates for a source/sink to a specific replica allows any derived relations such as `mz_{source|sink}_statuses` to make determinations based on all the replicas in the cluster.
+
+### Rollout strategy
+
+There is opportunity for the above work to be broken down into a couple of smaller chunks. The PRs involved may look like the following:
+
+1. Shift status writes to the storage controller
+2. Update `mz_{source|sink}_status_history` and any derived relations to account for a `replica_id`, which is necessary for multi-replica storage clusters
+
+## Alternatives
+
+<!--
+Similar to the Description section. List of alternative approaches considered,
+pros/cons or why they were not chosen.
+-->
+
+1. Add a graceful shutdown mechanism for storage workers
+    - Rejected due to the fact that graceful shutdown is not guaranteed to occur
+
+## Open questions
+
+<!--
+Anything currently unanswered that needs specific focus. This section may be
+expanded during the doc meeting as other unknowns are pointed out. These
+questions may be technical, product, or anything in-between.
+-->
+
+- The issue thread contains a good comment about how to handle scenarios in which `environmentd` crashes before a status update is written to persist. It seems like maintaining a durable log of pending updates and then replaying them when the storage controller starts is a possible solution. While we perform rehydration upon storage replicas crashing, I'm not sure if machinery exists in the storage controller to do the same upon `environmentd` crashing. Does this sound correct, or is there something I'm missing?
```

Comment:
I don't think you need to log out the pending updates; you can infer what's missing when the storage controller starts up. On startup, the catalog will re-install in the storage controller all live sources, sinks, storage instances, and replicas, and then call `initialization_complete`. At that moment, the storage controller should look for any sources/sinks that belong to an instance with no replicas, and ensure that those sources/sinks are marked as paused. Does that make sense?

Reply:
Ahh, yes makes sense!